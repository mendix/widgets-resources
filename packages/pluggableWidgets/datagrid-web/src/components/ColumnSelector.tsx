import { createElement, Dispatch, ReactElement, SetStateAction, useRef, useState } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faEye } from "@fortawesome/free-regular-svg-icons";
import { useOnClickOutside } from "@mendix/piw-utils-internal";
import { ColumnProperty } from "./Table";

export interface ColumnSelectorProps {
    columns: ColumnProperty[];
    hiddenColumns: string[];
    setHiddenColumns: Dispatch<SetStateAction<string[]>>;
}

export function ColumnSelector(props: ColumnSelectorProps): ReactElement {
    const [show, setShow] = useState(false);
    const componentRef = useRef<HTMLDivElement>(null);
    useOnClickOutside(componentRef, () => setShow(false));
    return (
        <div className="th column-selector">
            <div ref={componentRef} className="column-selector-content">
                <button
                    className="btn btn-default column-selector-button"
                    onClick={() => {
                        setShow(show => !show);
                    }}
                >
                    <FontAwesomeIcon icon={faEye} />
                </button>
                {show && (
                    <ul className="column-selectors">
                        {props.columns.map((column: ColumnProperty, index: number) => {
                            const isVisible = !props.hiddenColumns.includes(column.id);
                            return column.canHide ? (
                                <li key={index}>
                                    <input
                                        checked={isVisible}
                                        disabled={isVisible && props.columns.length - props.hiddenColumns.length === 1}
                                        id={`checkbox_toggle_${index}`}
                                        onClick={() => {
                                            props.setHiddenColumns(prev => {
                                                if (!isVisible) {
                                                    prev.splice(
                                                        prev.findIndex(v => v === column.id),
                                                        1
                                                    );
                                                    return [...prev];
                                                } else {
                                                    return [...prev, column.id];
                                                }
                                            });
                                        }}
                                        type="checkbox"
                                    />
                                    <label htmlFor={`checkbox_toggle_${index}`}>{column.header}</label>
                                </li>
                            ) : null;
                        })}
                    </ul>
                )}
            </div>
        </div>
    );
}
