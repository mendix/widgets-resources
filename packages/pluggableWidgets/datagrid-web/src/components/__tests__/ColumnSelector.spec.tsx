import { shallow } from "enzyme";
import { createElement } from "react";
import { ColumnSelector, ColumnSelectorProps } from "../ColumnSelector";
import { ColumnInstance } from "react-table";

describe("Column Selector", () => {
    it("renders the structure correctly", () => {
        const component = shallow(<ColumnSelector {...mockColumnSelectorProps()} />);

        expect(component).toMatchSnapshot();
    });
});

function mockColumnSelectorProps(): ColumnSelectorProps<object> {
    return {
        allColumns: [
            {
                id: "id",
                Header: "Test",
                filter: "text",
                isVisible: false,
                canHide: false,
                canDrag: false,
                canResize: false,
                disableSortBy: false,
                disableFilters: false
            }
        ] as Array<ColumnInstance<object>>,
        setHiddenColumns: jest.fn()
    };
}
