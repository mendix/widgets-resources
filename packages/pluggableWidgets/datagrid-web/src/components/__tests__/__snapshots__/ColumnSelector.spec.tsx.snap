// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column Selector renders the structure correctly 1`] = `
<th
  className="th column-selector"
  style={
    Object {
      "width": "1px",
    }
  }
>
  <button
    className="btn btn-default"
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            576,
            512,
            Array [],
            "f06e",
            "M288 144a110.94 110.94 0 0 0-31.24 5 55.4 55.4 0 0 1 7.24 27 56 56 0 0 1-56 56 55.4 55.4 0 0 1-27-7.24A111.71 111.71 0 1 0 288 144zm284.52 97.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400c-98.65 0-189.09-55-237.93-144C98.91 167 189.34 112 288 112s189.09 55 237.93 144C477.1 345 386.66 400 288 400z",
          ],
          "iconName": "eye",
          "prefix": "far",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </button>
</th>
`;
